File: flatros2/flatros2/subscription.py
# Copyright 2025 Ekumen, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from types import TracebackType
from typing import Callable, Generic, Optional, Type, TypeVar, Union

from rclpy.callback_groups import CallbackGroup
from rclpy.node import Node
from rclpy.qos import QoSProfile
from rclpy.waitable import NumberOfEntities, Waitable
from rclpy.type_support import check_for_type_support
from rclpy.impl.implementation_singleton import rclpy_implementation

from flatros2.message import Flat

WaitSet = rclpy_implementation.WaitSet

MessageT = TypeVar("MessageT")

class FlatSubscription(Generic[MessageT], Waitable):

    def __init__(
        self,
        node: Node,
        topic_type: Type[MessageT],
        topic_name: str,
        callback: Callable[[MessageT], None],
        qos_profile: Union[int, QoSProfile],
        callback_group: Optional[CallbackGroup] = None,
    ):
        self.node = node
        if not isinstance(topic_type, Flat):
            topic_type = Flat(topic_type)
        self.topic_type = topic_type
        if callback_group is None:
            callback_group = self.node.default_callback_group
        super().__init__(callback_group)
        self.node._validate_topic_or_service_name(topic_name)
        topic_name = self.node.resolve_topic_name(topic_name)
        self.callback = callback
        if not isinstance(qos_profile, QoSProfile):
            qos_profile = QoSProfile(depth=qos_profile)
        self.qos_profile = qos_profile
        from .flatros2_pybindings import FlatSubscriptionImpl
        self.__sub = FlatSubscriptionImpl(
            self.node.handle, self.topic_type, topic_name,
            self.qos_profile.get_c_qos_profile()
        )
        self.node.add_waitable(self)

    @property
    def topic_name(self) -> str:
        with self.__sub:
            return self.__sub.get_topic_name()

    def destroy(self) -> None:
        self.node.remove_waitable(self)
        self.__sub.destroy_when_not_in_use()

    def __enter__(self) -> None:
        self.__sub.__enter__()

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]] = None,
        exc_val: Optional[BaseException] = None,
        exc_tb: Optional[TracebackType] = None,
    ) -> None:
        self.__sub.__exit__(exc_type, exc_val, exc_tb)

    def is_ready(self, wait_set: WaitSet) -> bool:
        with self.__sub, wait_set:
            return self.__sub.is_ready(wait_set)

    def take_data(self) -> MessageT:
        with self.__sub:
            return self.__sub.take_loaned_message()

    async def execute(self, taken_data: MessageT) -> None:
        """Execute work after data has been taken from a ready wait set."""
        if taken_data:
            with self.__sub:
                try:
                    self.callback(taken_data)
                finally:
                    return self.__sub.return_loaned_message(taken_data)

    def get_num_entities(self) -> NumberOfEntities:
        return NumberOfEntities(num_subs=1)

    def add_to_wait_set(self, wait_set: WaitSet) -> None:
        with self.__sub, wait_set:
            self.__sub.add_to_waitset(wait_set)

File: flatros2/src/pybind11/subscription.cpp
// Copyright 2025 Ekumen, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "subscription.hpp"

#include "typesupport.hpp"
#include "utilities.hpp"

#include <rmw/types.h>

#include <rcl/types.h>
#include <rcl/subscription.h>

#include <rosidl_runtime_c/message_type_support_struct.h>

#include <stdexcept>

namespace flatros2 {
namespace pybind11 {

FlatSubscriptionImpl::FlatSubscriptionImpl(
  py::object pynode_handle, py::object pymessage_type,
  const std::string &topic_name, py::object pyqos_profile)
: pynode_handle_{std::move(pynode_handle)},
  pymessage_type_{std::move(pymessage_type)},
  rcl_subscription_{rcl_get_zero_initialized_subscription()}
{
  pynode_handle_.attr("__enter__")();
  rcl_subscription_options_t subscription_options = rcl_subscription_get_default_options();
  subscription_options.qos = pyqos_profile.cast<rmw_qos_profile_t>();
  subscription_options.disable_loaned_message = false;
  rcl_ret_t ret = rcl_subscription_init(
    &rcl_subscription_, get_rcl_node(pynode_handle_),
    get_message_type_support(pymessage_type_),
    topic_name.c_str(), &subscription_options);
  if (RCL_RET_OK != ret) {
    pynode_handle_.attr("__exit__")(py::none(), py::none(), py::none());
    throw std::runtime_error("failed to initialize subscription");
  }
}

FlatSubscriptionImpl::~FlatSubscriptionImpl() { destroy(); }

void FlatSubscriptionImpl::destroy() {
  if (rcl_subscription_is_valid(&rcl_subscription_)) {
    rcl_ret_t ret = rcl_subscription_fini(&rcl_subscription_, get_rcl_node(pynode_handle_));
    if (RCL_RET_OK != ret) {
      throw std::runtime_error("failed to finalize subscription");
    }
    pynode_handle_.attr("__exit__")(py::none(), py::none(), py::none());
  }
}

std::string FlatSubscriptionImpl::get_topic_name() const {
  const char * topic_name = rcl_subscription_get_topic_name(&rcl_subscription_);
  if (!topic_name) {
    throw std::runtime_error("failed to get subscription topic name");
  }
  return std::string(topic_name);
}

py::object FlatSubscriptionImpl::take_loaned_message() {
  if (!ready_) {
    return py::none();
  }
  ready_ = false;
  void * message = nullptr;
  rcl_ret_t ret = rcl_take_loaned_message(&rcl_subscription_, &message, nullptr, nullptr);
  if (RCL_RET_OK != ret) {
    if (RCL_RET_BAD_ALLOC == ret) {
      throw std::bad_alloc();
    }
    if (RCL_RET_SUBSCRIPTION_TAKE_FAILED == ret) {
      return py::none();
    }
    if (RCL_RET_OK != ret) {
      throw std::runtime_error("failed to take loaned message");
    }
  }
  return wrap_loaned_message(message, pymessage_type_);
}

void FlatSubscriptionImpl::return_loaned_message(py::object pymessage) {
  void * message = unwrap_loaned_message(pymessage, pymessage_type_);
  rcl_ret_t ret = rcl_return_loaned_message_from_subscription(&rcl_subscription_, message);
  if (RCL_RET_OK != ret) {
    throw std::runtime_error("failed to return loaned message");
  }
}

void FlatSubscriptionImpl::add_to_waitset(py::object pywait_set) {
  auto *wait_set =
    reinterpret_cast<rcl_wait_set_t *>(pywait_set.attr("pointer").cast<size_t>());
  rcl_ret_t ret = rcl_wait_set_add_subscription(wait_set, &rcl_subscription_, &index_);
  if (RCL_RET_OK != ret) {
    throw std::runtime_error("failed to add subscription to wait set");
  }
}

bool FlatSubscriptionImpl::is_ready(py::object pywait_set) {
  const auto *wait_set =
    reinterpret_cast<const rcl_wait_set_t *>(pywait_set.attr("pointer").cast<size_t>());
  if (index_ >= wait_set->size_of_subscriptions) {
    throw std::out_of_range("wait set index too big");
  }
  ready_ = (wait_set->subscriptions[index_] != nullptr);
  return ready_;
}

}  // namespace pybind11
}  // namespace flatros2
